File: MobMuPlatEditor-OSXLionAndUp/MobMuPlatEditor/MMPMultiTouch.m

//
//  MMPMultiTouch.m
//  MobMuPlatEditor
//
//  Created by diglesia on 2/24/14.
//  Copyright (c) 2014 Daniel Iglesia. All rights reserved.
//

#import "MMPMultiTouch.h"
#define BORDER_WIDTH 3
#define CURSOR_WIDTH 2
#define TOUCH_VIEW_RADIUS 20


@interface MMPMultiTouch () {
  NSMutableArray* _touchStack;
  NSMutableArray* _touchByVoxArray;//add, then hold NSNull values for empty voices
  
  NSView* borderView;
  TouchViewGroup* _currTouchViewGroup;
  BOOL _createdTouchOnMouseDown;
  MyTouch* _currMyTouch;
}

@end

@implementation MMPMultiTouch

- (id)initWithFrame:(CGRect)frame
{
  self = [super initWithFrame:frame];
  if (self) {
    self.address=@"/myMultiTouch";
    //self.backgroundColor = [UIColor purpleColor];
    //[self setMultipleTouchEnabled:YES];
    //self.clipsToBounds = YES;
    borderView = [[NSView alloc]init];
    [borderView setWantsLayer:YES];
    borderView.layer.borderWidth = BORDER_WIDTH;
    borderView.layer.borderColor = self.color.CGColor;
    [self addSubview:borderView];

    
    _touchStack = [[NSMutableArray alloc] init];
    _touchByVoxArray = [[NSMutableArray alloc] init];
    //[self setFrame:frame];
    
    [self addHandles];
    [self resizeSubviewsWithOldSize:self.frame.size];
  }
  return self;
}

-(void)hackRefresh{
  [super hackRefresh];
  borderView.layer.borderWidth = BORDER_WIDTH;
  //todo clear mock touches
}

- (void)resizeSubviewsWithOldSize:(NSSize)oldBoundsSize{
  [super resizeSubviewsWithOldSize:oldBoundsSize];


  [borderView setFrame:CGRectMake(0, 0, self.frame.size.width, self.frame.size.height)];
  [self removeTouches:[_touchStack copy]];//copy becase remove touches send s list of touches to remove from touchstack...
}

-(void)setColor:(NSColor *)inColor{
  [super setColor:inColor];
  borderView.layer.borderColor=inColor.CGColor;

}

//clip touch views to bounds
-(CGPoint)clipPoint:(CGPoint)inPoint{
  CGPoint outPoint;
  outPoint.x = MIN(MAX(0, inPoint.x), self.frame.size.width);
  outPoint.y = MIN(MAX(0, inPoint.y), self.frame.size.height);
  return outPoint;
}

-(CGPoint)normAndClipPoint:(CGPoint)inPoint{
  CGPoint outPoint;
  outPoint.x = inPoint.x/self.frame.size.width;
  outPoint.x = MIN(1, MAX(0, outPoint.x));
  outPoint.y = 1.0-(inPoint.y/self.frame.size.height);
  outPoint.y = MIN(1, MAX(0, outPoint.y));
  return outPoint;
}

-(void)mouseDown:(NSEvent *)theEvent{
  [super mouseDown:theEvent];
  //NSLog(@"down view");
  if(![self.editingDelegate isEditing]){
    
    CGPoint hitTestPoint = [[self superview] convertPoint:[theEvent locationInWindow] fromView:nil];
    CGPoint localPoint = [self convertPoint:[theEvent locationInWindow] fromView:nil];
    CGPoint clippedPoint = [self clipPoint:localPoint];
    
    NSView* topView = [self hitTest:hitTestPoint];
    if(topView==borderView || [topView isMemberOfClass:[NSView class]]){//topview or cursors
      
      
      //Touchview
      TouchViewGroup* tvg = [[TouchViewGroup alloc]init];
      tvg.touchView = [[TouchView alloc] initWithFrame:CGRectMake(clippedPoint.x-TOUCH_VIEW_RADIUS, clippedPoint.y-TOUCH_VIEW_RADIUS, TOUCH_VIEW_RADIUS*2, TOUCH_VIEW_RADIUS*2)];
      tvg.touchView.myGroup = tvg;
      [tvg.touchView setWantsLayer:YES];
      //tv.layer.backgroundColor = [self.highlightColor CGColor];
      tvg.touchView.layer.borderColor = [[NSColor blackColor] CGColor];
      tvg.touchView.layer.borderWidth = CURSOR_WIDTH;
      tvg.touchView.layer.cornerRadius = TOUCH_VIEW_RADIUS;
      
      tvg.cursorX = [[NSView alloc]init]; //]WithFrame:CGRectMake(5, 5, self.frame.size.width, CURSOR_WIDTH)];
      tvg.cursorY = [[NSView alloc]init]; //WithFrame:CGRectMake(5, 5, CURSOR_WIDTH, self.frame.size.height)];
      [tvg.cursorX setWantsLayer:YES];
      [tvg.cursorY setWantsLayer:YES];
      tvg.cursorX.layer.backgroundColor = [self.highlightColor CGColor];
      tvg.cursorY.layer.backgroundColor = [self.highlightColor CGColor];
      
      CGRect horizFrame = CGRectMake(0, clippedPoint.y-CURSOR_WIDTH/2, self.frame.size.width, CURSOR_WIDTH);
      [tvg.cursorX setFrame:horizFrame];
      CGRect vertFrame = CGRectMake(clippedPoint.x-CURSOR_WIDTH/2, 0, CURSOR_WIDTH, self.frame.size.height);
      [tvg.cursorY setFrame:vertFrame];
      
      [self addSubview:tvg.cursorX];
      [self addSubview:tvg.cursorY];
      [self addSubview:tvg.touchView];
      
      _currTouchViewGroup = tvg;
      _createdTouchOnMouseDown = YES;
      
      //stack
      MyTouch* myTouch = [[MyTouch alloc]init];
      myTouch.point = [self normAndClipPoint:localPoint];
      //myTouch.origEvent = theEvent;
      //myTouch.state = 1;
      
      myTouch.touchViewGroup = tvg;
      tvg.myTouch = myTouch;
      _currMyTouch = myTouch;
      
      [_touchStack addObject:myTouch];
      
      
      
      
      
      //poly vox
      BOOL added=NO;
      for(id element in _touchByVoxArray){//find NSNull vox slot
        if(element == [NSNull null]){
          int index = [_touchByVoxArray indexOfObject:element];
          myTouch.polyVox = index+1;
          [_touchByVoxArray replaceObjectAtIndex:index withObject:myTouch];
          added=YES;
          break;
        }
      }
      if(!added){
        [_touchByVoxArray addObject:myTouch];//add to end
        int index = (int)[_touchByVoxArray indexOfObject:myTouch];//or just array size//todo make polyvox nsuinteger
        myTouch.polyVox = index+1;
      }
      
      
      NSMutableArray* formattedMessageArray = [[NSMutableArray alloc]init];
      [formattedMessageArray addObject:self.address];
      [formattedMessageArray addObject:@"touch"];
      [formattedMessageArray addObject:[NSNumber numberWithInt:myTouch.polyVox]];
      [formattedMessageArray addObject:[NSNumber numberWithInt:1]];
      [formattedMessageArray addObject:[NSNumber numberWithFloat:myTouch.point.x]];
      [formattedMessageArray addObject:[NSNumber numberWithFloat:myTouch.point.y]];
      [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
      
      
      [self sendState];
      
      if([_touchStack count]>0) borderView.layer.borderColor= self.highlightColor.CGColor;
      //[self redrawCursors];
  
    }
    else if ([topView isKindOfClass:[TouchView class]]){
      _currTouchViewGroup = ((TouchView*)topView).myGroup;
      _currMyTouch = _currTouchViewGroup.myTouch;
      _createdTouchOnMouseDown=NO;
    }
    else {
      _currTouchViewGroup = nil;
      _createdTouchOnMouseDown=NO;
    }
  }
  
}

-(void)mouseDragged:(NSEvent *)theEvent{
  [super mouseDragged:theEvent];
  //NSLog(@"drag view");
  if(![self.editingDelegate isEditing]){
    
    //touchview
    CGPoint point = [self convertPoint:[theEvent locationInWindow] fromView:nil];
    CGPoint clippedPoint = [self clipPoint:point];
    
    [_currTouchViewGroup.touchView setFrame:CGRectMake(clippedPoint.x-TOUCH_VIEW_RADIUS, clippedPoint.y-TOUCH_VIEW_RADIUS, TOUCH_VIEW_RADIUS*2, TOUCH_VIEW_RADIUS*2)];
    
    CGRect horizFrame = CGRectMake(0, clippedPoint.y-CURSOR_WIDTH/2, self.frame.size.width, CURSOR_WIDTH);
    [_currTouchViewGroup.cursorX setFrame:horizFrame];
    CGRect vertFrame = CGRectMake(clippedPoint.x-CURSOR_WIDTH/2, 0, CURSOR_WIDTH, self.frame.size.height);
    [_currTouchViewGroup.cursorY setFrame:vertFrame];
    
    //for(MyTouch* myTouch in _touchStack){//TODO optimze! just get object by reference somehow? or use "indexOfObject"
      //if(myTouch.origEvent==theEvent){
        _currMyTouch.point = [self normAndClipPoint:point];
      
        //touchview
        //[_currTouchView setFrame:CGRectMake(point.x-TouchViewRadius, point.y-TouchViewRadius, TouchViewRadius*2, TouchViewRadius*2)];
     
        NSMutableArray* formattedMessageArray = [[NSMutableArray alloc]init];
        [formattedMessageArray addObject:self.address];
        [formattedMessageArray addObject:@"touch"];
        [formattedMessageArray addObject:[NSNumber numberWithInt:_currMyTouch.polyVox]];
        [formattedMessageArray addObject:[NSNumber numberWithInt:2]];
        [formattedMessageArray addObject:[NSNumber numberWithFloat:_currMyTouch.point.x]];
        [formattedMessageArray addObject:[NSNumber numberWithFloat:_currMyTouch.point.y]];
        [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
     
    //}
  
  [self sendState];
  //[self redrawCursors];
  }
}

-(void)mouseUp:(NSEvent *)theEvent{
  [super mouseUp:theEvent];
  if(![self.editingDelegate isEditing]){
    
    //NSLog(@"click count %d", [theEvent clickCount]);
    if([theEvent clickCount]==1 && _createdTouchOnMouseDown==NO){
      //[_currTouchView removeFromSuperview];
    
      NSMutableArray* touchesToRemoveArray = [[NSMutableArray alloc] init];
      //for(MyTouch* myTouch in _touchStack){//TODO optimze! just get object by reference somehow? or use "indexOfObject"!
			//if(myTouch.origEvent==theEvent){
        CGPoint point = [self convertPoint:[theEvent locationInWindow] fromView:nil];
        _currMyTouch.point = [self normAndClipPoint:point];//necc??
        [touchesToRemoveArray addObject:_currMyTouch];
        //CGPoint currPoint= [touch locationInView:self];
        //NSLog(@"remove touches : %.2f %.2f", currPoint.x, currPoint.y);
      //}
      //}
      [self removeTouches:touchesToRemoveArray];
  
          }
  }
}

-(void)removeTouches:(NSArray*)touchesToRemoveArray{
  [_touchStack removeObjectsInArray:touchesToRemoveArray];
  
  for(MyTouch* myTouch in touchesToRemoveArray){
    [myTouch.touchViewGroup.touchView removeFromSuperview];
    [myTouch.touchViewGroup.cursorX removeFromSuperview];
    [myTouch.touchViewGroup.cursorY removeFromSuperview];
    
    NSMutableArray* formattedMessageArray = [[NSMutableArray alloc]init];
    [formattedMessageArray addObject:self.address];
    [formattedMessageArray addObject:@"touch"];
    [formattedMessageArray addObject:[NSNumber numberWithInt:myTouch.polyVox]];
    [formattedMessageArray addObject:[NSNumber numberWithInt:0]];
    [formattedMessageArray addObject:[NSNumber numberWithFloat:myTouch.point.x]];
    [formattedMessageArray addObject:[NSNumber numberWithFloat:myTouch.point.y]];
    [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
    
    
    [_touchByVoxArray replaceObjectAtIndex:[_touchByVoxArray indexOfObject:myTouch] withObject:[NSNull null]];
  }
  
  [self sendState];
  if([_touchStack count]==0) borderView.layer.borderColor=self.color.CGColor;
  //[self redrawCursors];

}



-(void)sendState{
  //all are triplets
  //send stack as is:"/touchesByTime"
  //send by vox with nulls removed:
  //send prev sorted by x,y
  
  NSMutableArray* valArray = [NSMutableArray arrayWithArray:_touchStack];
  /*for(UITouch* touch in _touchStack){
   CGPoint currPoint= [touch locationInView:self];
   [valArray addObject:[NSNumber numberWithInt:[_touchByVoxArray indexOfObject:touch]+1]];//float or int?
   [valArray addObject:[NSNumber numberWithFloat:currPoint.x]];
   [valArray addObject:[NSNumber numberWithFloat:currPoint.y]];
   }*/
  
  //send as is
  NSMutableArray* msgArray = [[NSMutableArray alloc]init];
  [msgArray addObject:self.address];
  [msgArray addObject:@"touchesByTime"];
  for(MyTouch* myTouch in valArray){
    [msgArray addObject:[NSNumber numberWithInt:myTouch.polyVox]];//float or int?
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.x]];
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.y]];
  }
  [self.editingDelegate sendFormattedMessageArray:msgArray];
  
  //sort via vox
  [valArray sortUsingComparator:^NSComparisonResult(MyTouch* myTouch1, MyTouch* myTouch2){
    if(myTouch1.polyVox < myTouch2.polyVox) return NSOrderedAscending;
    else if (myTouch1.polyVox > myTouch2.polyVox) return NSOrderedDescending;
    else return NSOrderedSame;
  }];
  
  [msgArray removeAllObjects];
  [msgArray addObject:self.address];
  [msgArray addObject:@"touchesByVox"];
  for(MyTouch* myTouch in valArray){
    [msgArray addObject:[NSNumber numberWithInt:myTouch.polyVox]];//float or int?
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.x]];
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.y]];
  }
  [self.editingDelegate sendFormattedMessageArray:msgArray];
  
  //sort via X
  [valArray sortUsingComparator:^NSComparisonResult(MyTouch* myTouch1, MyTouch* myTouch2){
    
    if(myTouch1.point.x < myTouch2.point.x) return NSOrderedAscending;
    else if (myTouch1.point.x > myTouch2.point.x) return NSOrderedDescending;
    else return NSOrderedSame;
  }];
  
  [msgArray removeAllObjects];
  [msgArray addObject:self.address];
  [msgArray addObject:@"touchesByX"];
  
  for(MyTouch* myTouch in valArray){
    [msgArray addObject:[NSNumber numberWithInt:myTouch.polyVox]];//float or int?
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.x]];
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.y]];
  }
  [self.editingDelegate sendFormattedMessageArray:msgArray];
  
  //sort via Y
  [valArray sortUsingComparator:^NSComparisonResult(MyTouch* myTouch1, MyTouch* myTouch2){
    
    if(myTouch1.point.y < myTouch2.point.y) return NSOrderedAscending;
    else if (myTouch1.point.y > myTouch2.point.y) return NSOrderedDescending;
    else return NSOrderedSame;
  }];
  
  [msgArray removeAllObjects];
  [msgArray addObject:self.address];
  [msgArray addObject:@"touchesByY"];
  for(MyTouch* myTouch in valArray){
    [msgArray addObject:[NSNumber numberWithInt:myTouch.polyVox]];//float or int?
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.x]];
    [msgArray addObject:[NSNumber numberWithFloat:myTouch.point.y]];
  }
  [self.editingDelegate sendFormattedMessageArray:msgArray];
  
}


@end


@implementation MyTouch

@end

@implementation TouchViewGroup

@end

@implementation TouchView

@end

File: MobMuPlatEditor-Java/src/com/iglesiaintermedia/MobMuPlatEditor/controls/MMPMultiTouch.java
package com.iglesiaintermedia.MobMuPlatEditor.controls;

import java.awt.Color;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.event.MouseEvent;
import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import javax.swing.BorderFactory;
import javax.swing.JPanel;

public class MMPMultiTouch extends MMPControl{
	static final int BORDER_WIDTH = 3;
	static final int TOUCH_VIEW_RADIUS = 20;
	static final int CURSOR_WIDTH = 2;

	ArrayList<MyTouch> _touchStack;
	ArrayList<MyTouch> _touchByVoxArray;//add, then hold NSNull values for empty voices
	  
	JPanel borderPanel;
	TouchViewGroup _currTouchViewGroup;
	boolean _createdTouchOnMouseDown;
	MyTouch _currMyTouch;
	  
	public MMPMultiTouch(MMPMultiTouch otherMT){
		this(otherMT.getBounds());//normal constructor
		this.setColor(otherMT.color);
		this.setHighlightColor(otherMT.highlightColor);
		this.address=otherMT.address;
	}
	
	public MMPMultiTouch(Rectangle frame){
		super();
		address="/myMultiTouch";
		
		borderPanel = new JPanel();
		borderPanel.setOpaque(false);
		this.add(borderPanel);

	    _touchStack = new ArrayList<MyTouch>();
	    _touchByVoxArray = new ArrayList<MyTouch>();
		
		this.addMouseListener(this);
		this.addMouseMotionListener(this);
		this.setColor(color);
		this.setBounds(frame);
	}
	
	public void setBounds(Rectangle frame){
		super.setBounds(frame);
		borderPanel.setBounds(0,0,this.getWidth(), this.getHeight());
		//removeTouches(_touchStack);TODO
	}
	
	public void setColor(Color inColor){
		super.setColor(inColor);
		borderPanel.setBorder(BorderFactory.createLineBorder(inColor, BORDER_WIDTH));
		this.repaint();
	}
	
	Point clipPoint(Point inPoint) {
		Point outPoint = new Point();
		outPoint.x = Math.min(Math.max(0, inPoint.x), this.getWidth());
		outPoint.y = Math.min(Math.max(0, inPoint.y), this.getHeight());
		return outPoint;
	}
	
	Point2D.Float normAndClipPoint(Point inPoint){
		Point2D.Float outPoint = new Point2D.Float();
		outPoint.x = (float)inPoint.x/this.getWidth();
		outPoint.x = Math.min(1, Math.max(0, outPoint.x));
		outPoint.y = 1-((float)inPoint.y/this.getHeight());
		outPoint.y = Math.min(1, Math.max(0, outPoint.y));
		//System.out.print("norm in "+inPoint.x+" "+inPoint.y+" out "+outPoint.x+" "+outPoint.y);
		return outPoint;
	}
	
	public void mousePressed(MouseEvent e) {
		super.mousePressed(e);
		
		if(!editingDelegate.isEditing()){
			Point localPoint = e.getPoint();
			Point clippedPoint = clipPoint(localPoint);
			//scan trough touch views. WASTEFUL
			TouchView touchedView = null;
			for(MyTouch myTouch : _touchStack ){
				if(myTouch.touchViewGroup.touchView.getBounds().contains(localPoint) ) {
					//System.out.print("HIT");
					touchedView = myTouch.touchViewGroup.touchView;
				}
			}

			if(touchedView==null){
				//TouchView
				TouchViewGroup tvg = new TouchViewGroup();
				tvg.touchView = new TouchView();
				tvg.touchView.myGroup = tvg;
				tvg.touchView.setBounds(clippedPoint.x-TOUCH_VIEW_RADIUS, clippedPoint.y-TOUCH_VIEW_RADIUS, TOUCH_VIEW_RADIUS*2, TOUCH_VIEW_RADIUS*2);
				tvg.touchView.setCornerRadius(TOUCH_VIEW_RADIUS);
				tvg.touchView.setBackground(Color.BLACK);
				
				tvg.cursorX = new JPanel();
				tvg.cursorX.setBounds(0, clippedPoint.y-CURSOR_WIDTH/2, getWidth(), CURSOR_WIDTH);
				tvg.cursorX.setBackground(this.highlightColor);
				this.add(tvg.cursorX);
				
				tvg.cursorY = new JPanel();
				tvg.cursorY.setBounds(clippedPoint.x-CURSOR_WIDTH/2, 0, CURSOR_WIDTH, getHeight());
				tvg.cursorY.setBackground(this.highlightColor);
				this.add(tvg.cursorY);
				
				this.add(tvg.touchView);
				this.add(tvg.cursorX);
				this.add(tvg.cursorY);
				

				_currTouchViewGroup = tvg;
				_createdTouchOnMouseDown = true;

				//mytouch stack
				MyTouch myTouch = new MyTouch();
				myTouch.point = normAndClipPoint(e.getPoint());

				myTouch.touchViewGroup = tvg;
				tvg.myTouch = myTouch;
				_currMyTouch = myTouch;

				_touchStack.add(myTouch);


				//poly vox
				boolean added = false;
				for(Object element : _touchByVoxArray){
					if (element == null) {
						int index = _touchByVoxArray.indexOf(element);
						myTouch.polyVox = index + 1;
						_touchByVoxArray.set(index, myTouch);
						added = true;
						break;
					}
				}
				if (!added) {
					_touchByVoxArray.add(myTouch);
					int index = _touchByVoxArray.indexOf(myTouch);
					myTouch.polyVox = index + 1;
				}

				Object[] args = new Object[]{
						new String("touch"), 
						new Integer(myTouch.polyVox),
						new Integer(1),
						new Float(myTouch.point.x),
						new Float(myTouch.point.y)
				};
				editingDelegate.sendMessage(address, args);

				sendState();

				if(_touchStack.size()>0)
					borderPanel.setBorder(BorderFactory.createLineBorder(highlightColor, BORDER_WIDTH)); 
				

			}
			else {
				_currTouchViewGroup = touchedView.myGroup;
				_currMyTouch = _currTouchViewGroup.myTouch;
				_createdTouchOnMouseDown = false;
			}
		}
	}
	
	public void mouseDragged(MouseEvent e) {
		super.mouseDragged(e);
		
		if(!editingDelegate.isEditing()){
			Point clippedPoint = clipPoint(e.getPoint());
			
			_currTouchViewGroup.touchView.setBounds(clippedPoint.x-TOUCH_VIEW_RADIUS, clippedPoint.y-TOUCH_VIEW_RADIUS, TOUCH_VIEW_RADIUS*2, TOUCH_VIEW_RADIUS*2);
			_currTouchViewGroup.cursorX.setBounds(0, clippedPoint.y-CURSOR_WIDTH/2, getWidth(), CURSOR_WIDTH);
			_currTouchViewGroup.cursorY.setBounds(clippedPoint.x-CURSOR_WIDTH/2, 0, CURSOR_WIDTH, getHeight());
			_currMyTouch.point = normAndClipPoint(e.getPoint());
			
			Object[] args = new Object[]{
	    			new String("touch"), 
	    			new Integer(_currMyTouch.polyVox),
	    			new Integer(2),
	    			new Float(_currMyTouch.point.x),
	    			new Float(_currMyTouch.point.y)
	    	};
			editingDelegate.sendMessage(address, args);
			sendState();
		}
	}
	
	public void mouseReleased(MouseEvent e) {
		super.mouseReleased(e);
		//System.out.print("\nreleased");
	}
	
    public void mouseClicked(MouseEvent e) {
    	super.mouseClicked(e);
    	//System.out.print("\nclicked");
		if(!editingDelegate.isEditing()){
			if(_createdTouchOnMouseDown==false){
				ArrayList<MyTouch> touchesToRemoveArray = new ArrayList<MyTouch>();
				_currMyTouch.point = normAndClipPoint(e.getPoint());//necc?
				touchesToRemoveArray.add(_currMyTouch);
				removeTouches(touchesToRemoveArray);
				if (_touchStack.size()==0) borderPanel.setBorder(BorderFactory.createLineBorder(color, BORDER_WIDTH)); 
			}
		}	
	}
    
    void removeTouches(ArrayList<MyTouch> touchesToRemove) {
    	for(MyTouch myTouch : touchesToRemove) {
    		_touchStack.remove(myTouch);
    	
    		this.remove(myTouch.touchViewGroup.touchView);
    		this.remove(myTouch.touchViewGroup.cursorX);
    		this.remove(myTouch.touchViewGroup.cursorY);
    		
    		Object[] args = new Object[]{
	    			new String("touch"), 
	    			new Integer(myTouch.polyVox),
	    			new Integer(0),
	    			new Float(myTouch.point.x),
	    			new Float(myTouch.point.y)
	    	};
    		editingDelegate.sendMessage(address, args);
    		
    		_touchByVoxArray.set(_touchByVoxArray.indexOf(myTouch), null);
    	}
    	sendState();
    	borderPanel.repaint();
    	
    }
    
    void sendState(){
    	
    	ArrayList<MyTouch> valArray = new ArrayList<MyTouch>(_touchStack.size());
    	for(MyTouch touch : _touchStack){
    		valArray.add(touch.clone());
    	}
    	
    	//send as is.
    	ArrayList<Object> msgArray = new ArrayList<Object>(3*valArray.size()+1);
    	msgArray.add(new String("touchesByTime"));
    	for(MyTouch touch : valArray){
    		msgArray.add(new Integer(touch.polyVox));
    		msgArray.add(new Float(touch.point.x));
    		msgArray.add(new Float(touch.point.y));
    	}
    	editingDelegate.sendMessage(address, msgArray.toArray());
    	
    	//sort via vox
    	Collections.sort(valArray, new Comparator<MyTouch>() {
    	    public int compare(MyTouch a, MyTouch b) {
    	    	if(a.polyVox < b.polyVox) return -1;
    	        else if (a.polyVox > b.polyVox) return 1;
    	        else return 0;
    	    }
    	});
    	
    	msgArray.clear();
    	msgArray.add(new String("touchesByVox"));
    	for(MyTouch touch : valArray){
    		msgArray.add(new Integer(touch.polyVox));
    		msgArray.add(new Float(touch.point.x));
    		msgArray.add(new Float(touch.point.y));
    	}
    	editingDelegate.sendMessage(address, msgArray.toArray());
    	
    	//sort via X
    	Collections.sort(valArray, new Comparator<MyTouch>() {
    	    public int compare(MyTouch a, MyTouch b) {
    	    	if(a.point.x < b.point.x) return -1;
    	        else if (a.point.x > b.point.x) return 1;
    	        else return 0;
    	    }
    	});
    	
    	msgArray.clear();
    	msgArray.add(new String("touchesByX"));
    	for(MyTouch touch : valArray){
    		msgArray.add(new Integer(touch.polyVox));
    		msgArray.add(new Float(touch.point.x));
    		msgArray.add(new Float(touch.point.y));
    	}
    	editingDelegate.sendMessage(address, msgArray.toArray());
    	
    	//sort via Y
    	Collections.sort(valArray, new Comparator<MyTouch>() {
    	    public int compare(MyTouch a, MyTouch b) {
    	    	if(a.point.y < b.point.y) return -1;
    	        else if (a.point.y > b.point.y) return 1;
    	        else return 0;
    	    }
    	});
    	
    	msgArray.clear();
    	msgArray.add(new String("touchesByY"));
    	for(MyTouch touch : valArray){
    		msgArray.add(new Integer(touch.polyVox));
    		msgArray.add(new Float(touch.point.x));
    		msgArray.add(new Float(touch.point.y));
    	}
    	editingDelegate.sendMessage(address, msgArray.toArray());
    	
    }
	
}

class MyTouch{
	Point2D.Float point;
	TouchViewGroup touchViewGroup;
	int polyVox;
	
	public MyTouch clone(){
		MyTouch newTouch = new MyTouch();
		newTouch.point = (Point2D.Float)this.point.clone();
		newTouch.polyVox = this.polyVox;
		return newTouch;
	}
}

class TouchView extends RoundedPanel{
	TouchViewGroup myGroup;
}

class TouchViewGroup{
	MyTouch myTouch;
	TouchView touchView;
	JPanel cursorX;
	JPanel cursorY;
	
	
}


